miner/multi_collator.go

```diff
@@ -0,0 +1,394 @@
+	return new(big.Int).Set(bs.work.env.header.BaseFee)
+}
+
+func (bs *collatorBlockState) Signer() types.Signer {
+	return bs.work.env.signer
+}
+
+// BlockCollator is the publicly-exposed interface
+// for implementing custom block collation strategies
+type BlockCollator interface {
+	CollateBlock(bs BlockState, pool Pool)
+	/*
+	   // TODO implement these
+	   SideChainHook(header *types.Header)
+	   NewHeadHook(header *types.Header)
+	*/
+}
+
+type collator struct {
+	newWorkCh    chan collatorWork
+	workResultCh chan collatorWork
+	// channel signalling collator loop should exit
+	exitCh            chan struct{}
+	newHeadCh         chan types.Header
+	sideChainCh       chan types.Header
+	blockCollatorImpl BlockCollator
+
+	chainConfig *params.ChainConfig
+	chain       *core.BlockChain
+	pool        Pool
+}
+
+// each active collator runs mainLoop() in a goroutine.
+// It receives new work from the miner and listens for new blocks built from CollateBlock
+// calling Commit() on the provided collatorBlockState
+func (c *collator) mainLoop() {
+	for {
+		select {
+		case newWork := <-c.newWorkCh:
+			c.blockCollatorImpl.CollateBlock(&collatorBlockState{work: newWork, c: c, done: false}, c.pool)
+			// signal to the exitCh that the collator is done
+			// computing this work.
+			c.workResultCh <- collatorWork{env: nil, interrupt: nil, counter: newWork.counter}
+		case <-c.exitCh:
+			// TODO any cleanup needed?
+			return
+		case newHead := <-c.newHeadCh:
+			// TODO call hook here
+			_ = newHead
+		case sideHeader := <-c.sideChainCh:
+			_ = sideHeader
+			// TODO call hook here
+		default:
+		}
+	}
+}
+
+var (
+	// collator
+	workResultChSize = 10
+	newWorkChSize    = 10
+	newHeadChSize    = 10
+	sideChainChSize  = 10
+)
+
+// MultiCollator manages multiple active collators
+type MultiCollator struct {
+	counter                 uint64
+	responsiveCollatorCount int
+	collators               []collator
+	pool                    Pool
+	interrupt               *int32
+}
+
+func NewMultiCollator(chainConfig *params.ChainConfig, chain *core.BlockChain, pool Pool, strategies []BlockCollator) MultiCollator {
+	collators := []collator{}
+	for _, s := range strategies {
+		collators = append(collators, collator{
+			newWorkCh:         make(chan collatorWork, newWorkChSize),
+			workResultCh:      make(chan collatorWork, workResultChSize),
+			exitCh:            make(chan struct{}),
+			newHeadCh:         make(chan types.Header, newHeadChSize),
+			blockCollatorImpl: s,
+			chainConfig:       chainConfig,
+			chain:             chain,
+			pool:              pool,
+		})
+	}
+	return MultiCollator{
+		counter:                 0,
+		responsiveCollatorCount: 0,
+		collators:               collators,
+		interrupt:               nil,
+	}
+}
+
+func (m *MultiCollator) Start() {
+	for _, c := range m.collators {
+		go c.mainLoop()
+	}
+}
+
+func (m *MultiCollator) Close() {
+	for _, c := range m.collators {
+		select {
+		case c.exitCh <- struct{}{}:
```

Comment:
For exit channels, what usually works pretty good is just to close them when it's time to exit. Then it doesn't matter if there's one listening or zero or 100. Whereas if you send a `struct{}{}`, then you need exactly one listener -- or, do as you've done here, and have a default-case.

So unless you have some compelling reason, just `close(c.exitCh)` IMO